File: frontend/lib/types/festival.ts
/**
 * Festival-related TypeScript types
 *
 * These types match the backend API response structures
 * for festival endpoints.
 */

export type FestivalStatus = 'announced' | 'confirmed' | 'cancelled' | 'completed'

/** Labels for display */
export const FESTIVAL_STATUS_LABELS: Record<FestivalStatus, string> = {
  announced: 'Announced',
  confirmed: 'Confirmed',
  cancelled: 'Cancelled',
  completed: 'Completed',
}

/** All valid festival statuses for filter dropdowns */
export const FESTIVAL_STATUSES: FestivalStatus[] = [
  'announced',
  'confirmed',
  'cancelled',
  'completed',
]

/** Badge variant mapping for festival statuses */
export function getFestivalStatusVariant(
  status: string
): 'default' | 'secondary' | 'destructive' | 'outline' {
  switch (status) {
    case 'confirmed':
      return 'default'
    case 'announced':
      return 'secondary'
    case 'cancelled':
      return 'destructive'
    case 'completed':
      return 'outline'
    default:
      return 'secondary'
  }
}

/**
 * Get a display label for a festival status
 */
export function getFestivalStatusLabel(status: string): string {
  return (
    FESTIVAL_STATUS_LABELS[status as FestivalStatus] ??
    status.charAt(0).toUpperCase() + status.slice(1)
  )
}

export type BillingTier =
  | 'headliner'
  | 'sub_headliner'
  | 'mid_card'
  | 'undercard'
  | 'local'
  | 'dj'
  | 'host'

/** Labels for display */
export const BILLING_TIER_LABELS: Record<BillingTier, string> = {
  headliner: 'Headliner',
  sub_headliner: 'Sub-Headliner',
  mid_card: 'Mid Card',
  undercard: 'Undercard',
  local: 'Local',
  dj: 'DJ',
  host: 'Host',
}

/** All valid billing tiers for dropdowns */
export const BILLING_TIERS: BillingTier[] = [
  'headliner',
  'sub_headliner',
  'mid_card',
  'undercard',
  'local',
  'dj',
  'host',
]

/**
 * Get a display label for a billing tier
 */
export function getBillingTierLabel(tier: string): string {
  return (
    BILLING_TIER_LABELS[tier as BillingTier] ??
    tier
      .split('_')
      .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
      .join(' ')
  )
}

export interface FestivalSocial {
  instagram?: string | null
  facebook?: string | null
  twitter?: string | null
  youtube?: string | null
  spotify?: string | null
  website?: string | null
}

export interface FestivalDetail {
  id: number
  name: string
  slug: string
  series_slug: string
  edition_year: number
  description: string | null
  location_name: string | null
  city: string | null
  state: string | null
  country: string | null
  start_date: string
  end_date: string
  website: string | null
  ticket_url: string | null
  flyer_url: string | null
  status: string
  social: FestivalSocial | null
  artist_count: number
  venue_count: number
  created_at: string
  updated_at: string
}

export interface FestivalListItem {
  id: number
  name: string
  slug: string
  series_slug: string
  edition_year: number
  city: string | null
  state: string | null
  start_date: string
  end_date: string
  status: string
  artist_count: number
  venue_count: number
}

export interface FestivalsListResponse {
  festivals: FestivalListItem[]
  count: number
}

export interface FestivalArtist {
  id: number
  artist_id: number
  artist_slug: string
  artist_name: string
  billing_tier: string
  position: number
  day_date: string | null
  stage: string | null
  set_time: string | null
  venue_id: number | null
}

export interface FestivalArtistsResponse {
  artists: FestivalArtist[]
  count: number
}

export interface FestivalVenue {
  id: number
  venue_id: number
  venue_name: string
  venue_slug: string
  city: string
  state: string
  is_primary: boolean
}

export interface FestivalVenuesResponse {
  venues: FestivalVenue[]
  count: number
}

export interface ArtistFestivalListItem extends FestivalListItem {
  billing_tier: string
  day_date: string | null
  stage: string | null
}

export interface ArtistFestivalsResponse {
  festivals: ArtistFestivalListItem[]
  count: number
}

/**
 * Format a festival's location string
 */
export function formatFestivalLocation(festival: {
  city: string | null
  state: string | null
}): string | null {
  if (festival.city && festival.state) return `${festival.city}, ${festival.state}`
  if (festival.city) return festival.city
  if (festival.state) return festival.state
  return null
}

/**
 * Format a festival's date range string
 */
export function formatFestivalDates(startDate: string, endDate: string): string {
  try {
    // Parse as local dates (YYYY-MM-DD format, avoid timezone issues)
    const [startYear, startMonth, startDay] = startDate.split('-').map(Number)
    const [endYear, endMonth, endDay] = endDate.split('-').map(Number)
    const start = new Date(startYear, startMonth - 1, startDay)
    const end = new Date(endYear, endMonth - 1, endDay)

    const startStr = start.toLocaleDateString('en-US', {
      month: 'short',
      day: 'numeric',
    })
    const endStr = end.toLocaleDateString('en-US', {
      month: 'short',
      day: 'numeric',
      year: 'numeric',
    })
    return `${startStr} - ${endStr}`
  } catch {
    return `${startDate} - ${endDate}`
  }
}

File: backend/internal/models/catalog/festival.go
package catalog

import (
	"encoding/json"
	"time"
)

// FestivalStatus represents the current status of a festival
type FestivalStatus string

const (
	FestivalStatusAnnounced FestivalStatus = "announced"
	FestivalStatusConfirmed FestivalStatus = "confirmed"
	FestivalStatusCancelled FestivalStatus = "cancelled"
	FestivalStatusCompleted FestivalStatus = "completed"
)

// BillingTier represents the billing level of an artist at a festival
type BillingTier string

const (
	BillingTierHeadliner    BillingTier = "headliner"
	BillingTierSubHeadliner BillingTier = "sub_headliner"
	BillingTierMidCard      BillingTier = "mid_card"
	BillingTierUndercard    BillingTier = "undercard"
	BillingTierLocal        BillingTier = "local"
	BillingTierDJ           BillingTier = "dj"
	BillingTierHost         BillingTier = "host"
)

// Festival represents a music festival (distinct from a show — multi-day, tiered billing, multi-venue)
type Festival struct {
	ID           uint             `gorm:"primaryKey"`
	Name         string           `gorm:"not null"`
	Slug         string           `gorm:"not null;uniqueIndex"`
	SeriesSlug   string           `gorm:"column:series_slug;not null"`
	EditionYear  int              `gorm:"column:edition_year;not null"`
	Description  *string          `gorm:"column:description"`
	LocationName *string          `gorm:"column:location_name"`
	City         *string          `gorm:"column:city"`
	State        *string          `gorm:"column:state"`
	Country      *string          `gorm:"column:country"`
	StartDate    string           `gorm:"column:start_date;type:date;not null"`
	EndDate      string           `gorm:"column:end_date;type:date;not null"`
	Website      *string          `gorm:"column:website"`
	TicketURL    *string          `gorm:"column:ticket_url"`
	FlyerURL     *string          `gorm:"column:flyer_url"`
	Status       FestivalStatus   `gorm:"column:status;not null;default:'announced'"`
	Social       *json.RawMessage `gorm:"column:social;type:jsonb;default:'{}'"`

	// Data provenance fields
	DataSource       *string    `json:"data_source,omitempty" gorm:"column:data_source;size:50"`
	SourceConfidence *float64   `json:"source_confidence,omitempty" gorm:"column:source_confidence;type:numeric(3,2)"`
	LastVerifiedAt   *time.Time `json:"last_verified_at,omitempty" gorm:"column:last_verified_at"`

	CreatedAt time.Time `gorm:"not null"`
	UpdatedAt time.Time `gorm:"not null"`

	// Relationships
	Artists []FestivalArtist `gorm:"foreignKey:FestivalID"`
	Venues  []FestivalVenue  `gorm:"foreignKey:FestivalID"`
}

// TableName specifies the table name for Festival
func (Festival) TableName() string {
	return "festivals"
}

// FestivalArtist represents the junction table between festivals and artists
type FestivalArtist struct {
	ID          uint        `gorm:"primaryKey"`
	FestivalID  uint        `gorm:"column:festival_id;not null"`
	ArtistID    uint        `gorm:"column:artist_id;not null"`
	BillingTier BillingTier `gorm:"column:billing_tier;not null;default:'mid_card'"`
	Position    int         `gorm:"not null;default:0"`
	DayDate     *string     `gorm:"column:day_date;type:date"`
	Stage       *string     `gorm:"column:stage"`
	SetTime     *string     `gorm:"column:set_time;type:time"`
	VenueID     *uint       `gorm:"column:venue_id"`
	CreatedAt   time.Time   `gorm:"not null"`

	// Relationships
	Festival Festival `gorm:"foreignKey:FestivalID"`
	Artist   Artist   `gorm:"foreignKey:ArtistID"`
	Venue    *Venue   `gorm:"foreignKey:VenueID"`
}

// TableName specifies the table name for FestivalArtist
func (FestivalArtist) TableName() string {
	return "festival_artists"
}

// FestivalVenue represents the junction table between festivals and venues
type FestivalVenue struct {
	ID         uint      `gorm:"primaryKey"`
	FestivalID uint      `gorm:"column:festival_id;not null"`
	VenueID    uint      `gorm:"column:venue_id;not null"`
	IsPrimary  bool      `gorm:"column:is_primary;not null;default:false"`
	CreatedAt  time.Time `gorm:"not null"`

	// Relationships
	Festival Festival `gorm:"foreignKey:FestivalID"`
	Venue    Venue    `gorm:"foreignKey:VenueID"`
}

// TableName specifies the table name for FestivalVenue
func (FestivalVenue) TableName() string {
	return "festival_venues"
}
